lib/msf/core/opt_rhost_url.rb

```diff
@@ -0,0 +1,61 @@
+# -*- coding: binary -*-
+
+module Msf
+  ###
+  #
+  # RHOST URL option.
+  #
+  ###
+  class OptRhostUrl < OptBase
+    def type
+      'rhost url'
+    end
+
+
+    def normalize(value)
+      uri = URI(value)
+      option_hash = {}
+      # Blank this out since we don't know if this new value will have a `VHOST` to ensure we remove the old value
+      option_hash['VHOST'] = nil
+
+      option_hash['RHOSTS'] = uri.hostname
+      option_hash['RPORT'] = uri.port
+      option_hash['SSL'] = %w{ssl https}.include?(uri.scheme)
+
+      # Both `TARGETURI` and `URI` are used as datastore options to denote the path on a uri
+      option_hash['TARGETURI'] = uri.path
+      option_hash['URI'] = uri.path
+
+      if uri.scheme and %{http https}.include?(uri.scheme)
+        option_hash['VHOST'] = uri.hostname unless Rex::Socket.is_ip_addr?(uri.hostname)
+        option_hash['HttpUsername'] = uri.user.to_s
+        option_hash['HttpPassword'] = uri.password.to_s
+      end
+
+      option_hash
+    end
+
+    def valid?(value, check_empty: true)
+      uri = URI(value)
+      return false unless uri.host != nil and uri.port != nil
+      super
+    end
+
+    def calculate_value(datastore)
+      if datastore['RHOSTS']
+        begin
+          uri = URI::Generic.build(host: datastore['RHOSTS'])
+          uri.port=datastore['RPORT']
+          # The datastore uses both `TARGETURI` and `URI` to denote the path of a URL, we try both here and fall back to `/`
+          uri.path=(datastore['TARGETURI'] || datastore['URI'] || '/')
+          uri.user=datastore['HttpUsername']
+          uri.password=datastore['HttpPassword'] if uri.user
```

Comment:
Looks like the protocol's getting dropped at the minute, can we slice in the scheme too? 👀 

```
> set RHOST_URL http://google.com
RHOST_URL => //google.com:80
```

Reply:
Not in a reliable way that I was able to come up with no
see we aren't actually storing the scheme on our end, we'd have to sort of reverse engineer the scheme from the provided options but there's no guarantee we'd get it right, we could end up having the user set a url then immediately be showing them one with a different scheme

Reply:
I think what I could do though is slice off the leading `//` that would solve the problem I think

Reply:
It's an interesting discussion point. The original [RFC](https://github.com/rapid7/metasploit-framework/wiki/RFC---Metasploit-URL-support#additional-considerations) made notes about different schemes too - such as `ftp`, `postgres`, `smb`, etc.

Reply:
Unless you think it'll cause more confusion not having those formats work from the get go I still think this current iteration would benefit a majority of users and we can always add mroe use cases down the line when it becomes apparent they are warranted

Reply:
I'm just keeping an open mind for now, discussing how the different implementations could impact various workflows. We want to reduce the possibility of confusing users in the initial release, and avoiding potential confusion for when we do support more use cases in the future. Finding the right balance between usable, shippable, and iterable will be good :+1:

Reply:
I wonder if things become easier if this was renamed from the current generic name to `OptHttpUrl` ? 🤔

Reply:
Alan fights for the users.